Add Windows build

tkralphs opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

tkralphs committed:
Adding files for Windows build

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

tkralphs committed:
Reset build number

tkralphs commented:
@conda-forge-admin, please rerender

Commit:
MNT: Re-rendered with conda-build 25.1.2, conda-smithy 3.45.4, and conda-forge-pinning 2025.02.19.03.03.20

tkralphs commented:
@jpfeuffer, I took a stab at a windows build for Clp. We have the same issue here that arose in Osi---for some reason, Blas and Lapack are not getting linked, even though we provided the locations through `--with-blas-lib` and `--with-lapack-lib`. In the case of Osi, it arose when trying to link the unit test executable and we could ignore it. Here, we can't. So I guess I'll need to debug this a bit. But the good news is that it configured and got all the way to the link step for the executable on the first try. So we really should be able to get all projects done in relatively short order now.

jpfeuffer commented:
One question: should we be building static or dynamic libraries? From the artifacts produced it looks like they only include a static version (.lib file).
It also looks like your pkgconfig does not support static libraries because you would need to specify transitive requirements explicitly (which you don't). Only dynamic libraries know how to look up their own dependencies.

jpfeuffer commented:
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1180096&view=logs&j=fe7faf59-69bb-5487-b8fd-a35a7bda3a51&t=55731638-dd49-5144-2129-5101b9dd1960&l=1460

Yes it somehow thinks we want a static build.
How can we change that in the upstream packages?
It looks like for the other platforms it defaults to shared libs.

tkralphs commented:
Shared libraries are a pain on Windows (due to all the required `__declspec` stuff), so for a long time, we did not support producing DLLs and everything was linked statically. This is fixed in the `master` branches now, but this older version still needs to be statically linked. This shouldn't be an issue, it is working fine in our CI on Github, which builds using Actions. Support for pkg-config on Windows is definitely broken and shouldn't be used. It could be that we just need to explicitly disable it, although I guess that's not the issue either. We just don't get Lapack and Blas on the link line, even though we're explicitly specifying the location. I'll poke at it a bit.

tkralphs committed:
Disable pkg-config

jpfeuffer commented on recipe/build.sh:
```diff
@@ -19,15 +26,19 @@ fi
+  --with-blas-lib="${BLAS_LIB}" \
+  --with-lapack-lib="${LAPACK_LIB}" \
+  --disable-pkg-config \
+  ${EXTRA_FLAGS} || cat Clp/config.log
```
Maybe try to print the config.log always for now.
I saw some checks on your configure file that will test the user defined libs first. Maybe sth goes wrong there.

tkralphs committed:
Enable pkg-config again

tkralphs committed:
Remove dependency on lapack and blas

tkralphs commented:
OK, I think I may have figured out what is going on here. It's (most likely) an oversight (bug) in our autotools setup that is only problematic in the exact scenario that arises here. Weirdly, this combination of circumstances has never come up before. In a nutshell, it seems that because Blas and Lapack are not direct dependencies, the function in `configure.ac` that is required to make `--with-blas-lib` and `--with-lapack-lib` work properly was not being called. This is not an issue if any of the following are true:
- we have a non-Windows pkg-config (i.e., one that understands posix paths) installed and we are building with anything other than the Visual Studio compiler (again, posix paths),
- BLAS and LAPACK are available in a system directory where they can be discovered automatically, or
- we are building BLAS and LAPACK from source using our ThirdParty build mechanism.

In all the builds we're doing outside of conda-forge, we have one or more of these things being true on every platform, so this was never realized until now. Fixing this should be easy, but I need to make a new release of Clp (and I should also go back and do Osi, since this is why it's unit test doesn't work). This may take a bit of time.

One thing that is still very unclear to me, though, is how/where CoinUtils is finding the static BLAS and LAPACK libraries it is linking to. If you look at the link line for it's unit test, it contains
```
D:/bld/coin-or-utils_1740336896402/_h_env/Library/lib/lapack.lib D:/bld/coin-or-utils_1740336896402/_h_env/Library/lib/cblas.lib 
```
and these files seem to be found. But when I install the `coin-or-utils` package, I can't find those files in that location. And looking at the `json` file that lists what package are installed with `libcblas`, only a DLL is listed, not a static library (as I would expect). So it is a bit of a mystery to me how the CoinUtils unit test is building. 

As such, from what I see now, there is no way to actually link to the CoinUtils library successfully after installing the package. If we manage to build the `clp` and `cbc` executables, they will be statically linked and should work, but it doesn't looks to me like it will be possible to build, e.g., Python C extensions that link to these libraries, from source. 

All of this is fixed in `master` and it may make more sense to use what time I have to spend getting those out into release so we can just stop spinning our wheels on this. And of course, another alternative is just to give up on linking to Blas and Lapack for now and revisit when the new releases are available. Everything would still work, there would just be a performance hit. 

For now, I will fix this "bug" and make new releases. If that fixes it, then we can move on and I'll focus my available energies on getting the new releases out.

jpfeuffer commented:
Thank you for the investigation. Regarding cblas: since they are allowing for different implementations of blas, the actual headers and the corresponding link libraries (.lib) are in some kind of umbrella devel package. I already forgot the details but it was answered somewhere in my first PR to coin utils (and/or some question that I asked at the blas feedstock).

Regarding finding the libs here: can we maybe add a dir to be searched outside of your build system somehow by manipulating some env var (probably PATH on Windows)?

tkralphs commented:
>Regarding cblas: since they are allowing for different implementations of blas, the actual headers and the corresponding link libraries (.lib) are in some kind of umbrella devel package.

Yeah, I know the BLAS/LAPACK ecosystenm is complex and I've been going down that rabbit hole to try to understand. There are some meta-packages and then some packages with actual libraries. I can see the files installed with each package by looking at the json in the `conda-meta` directory. But the bottom line is that on the build machine on Azure, there is a file `<EnvironmentRoot>/Library/lib/cblas.lib`, since the linking succeeds with that file specified on the link line. It only makes sense that this file comes from the `cblas` package, but this doesn't seem to be the case. I simply cannot find any package containing the file `cblas.lib`. I've manualy installed all the packages that are specified in `meta.yml` and none install that file in `<EnvironmentRoot>/Library/lib/`.

When I search for some mention of a similar issue, the `gsl` package keeps coming up. I found an issue from someone who says the same thing. When they try to link to the `gsl` library (which depends on `cblas.lib`), the file `cblas.lib` cannot be found: https://github.com/conda-forge/gsl-feedstock/issues/51. This is what I also assume would happen if one tries to link to `libCoinUtils.lib`. 

There is a package `mkl-static`, which includes static libraries that I assume would work, but I'm hesitant to open another can of worms. I really have to climb out of this rabbit hole!

tkralphs commented:
> Regarding finding the libs here: can we maybe add a dir to be searched outside of your build system somehow by manipulating some env var (probably PATH on Windows)?

It would be possible to just link to some libraries manually installed on the system (not by conda/mamba). I don't think there's any need to specify a path to those in an environment variable if you want to go that route, just providing a full path on the link line is fine. THis is what's already done to link to `cblas.lib`. But getting the library from some manually installed library on the system seems to defeat the purpose of using conda/mamba for the install and would be confusing for the typical user. In any case, the files that apparently exist on the build machine are indeed installed somehow from within the conda/mamba ecosystem. There seems to be a way to do this somehow.

jpfeuffer commented:
The default implementation for building on conda-forge is libcblas=*=netlib (https://github.com/conda-forge/coin-or-clp-feedstock/pull/18/files#diff-283d4846cd6e9d12c8444a622943204d7a168d0c8c350127106345fd3f42ac2bR18).
Conda adds this pinning whenever you re-render your package when it encounters libcblas as a dependency.
libcblas netlib version [should have cblas.lib as a link library](https://conda-metadata-app.streamlit.app/?q=conda-forge%2Fwin-64%2Flibcblas-3.9.0-31_ha692739_openblas.conda). This is how I used it for coinutils.

Yes, I just mean to temporary alter the PATH during building. I thought the problem was that the fully specified "--with-blas-lib" does not get passed through to the link command in your build system. And I thought with the PATH trick would maybe solve this by circumventing the build system. Alternatively we could try to populate LINK_FLAGS or something.

tkralphs commented:
>libcblas netlib version [should have cblas.lib as a link library](https://conda-metadata-app.streamlit.app/?q=conda-forge%2Fwin-64%2Flibcblas-3.9.0-31_ha692739_openblas.conda). This is how I used it for coinutils.

Exactly, but only `cblas.dll` is listed as a provided file [here](https://conda-metadata-app.streamlit.app/?q=conda-forge%2Fwin-64%2Flibcblas-3.9.0-31_ha692739_openblas.conda), not `cblas.lib` and that's exactly what I see when I install it locally. Yet, somehow, `cblas.lib` is found on the build machine when building CoinUtils. It should also be there when building Clp and Clp will probably build correctly once I make the required change to our build system. But still, I don't see how one can use the libraries that are installed without having `cblas.lib` available. 

>I thought the problem was that the fully specified "--with-blas-lib" does not get passed through to the link command in your build system. 

Unfortunately, it's not a path issue, it's that Clp just ignores the `--with-blas-lib` and `--with-lapack-lib` altogether at the moment. So even if the libraries could be found within some standard path, the linking would not happen. The libraries just aren't added to the link line at all. With CoinUtils, they are added with a full explicit path and the linking is successful. The same should be the case for Clp, once I add this extra check.

tkralphs committed:
Explicitly specify the link line of CoinUtils

jpfeuffer commented:
> Exactly, but only `cblas.dll` is listed as a provided file [here](https://conda-metadata-app.streamlit.app/?q=conda-forge%2Fwin-64%2Flibcblas-3.9.0-31_ha692739_openblas.conda), not `cblas.lib` and that's exactly what I see when I install it locally.

Your link goes to the openblas version of cblas. I think only the netlib version has the link library because it is the default for **building** things on conda-forge. At runtime you could then allow for any of the cblas versions to be present (which provide only the dlls)

jpfeuffer commented:
I was thinking about something along the lines of the following variables, which should ignore your build system and automatically forward user flags to the link command line:
https://www.gnu.org/software/automake/manual/html_node/Libtool-Flags.html (at least this is the idea, as I am not too familiar with automake/conf)

tkralphs committed:
Try single quotes

tkralphs commented:
>Your link goes to the openblas version of cblas. I think only the netlib version has the link library because it is the default for building things on conda-forge. 

Are you sure about this? The library provided by openblas seems to be called `openblas.lib`, not `cblas.lib`. I also installed openblas and didn't get the library.

> At runtime you could then allow for any of the cblas versions to be present (which provide only the dlls)

We don't need DLLs because everything is statically linked. For using the `clp` executable, we don't need anything, since the libraries will be statically linked at build time. We only need libraries if we want to build another application that uses `libCoinUtils.lib`, such as a Python C extension. 

>I was thinking about something along the lines of the following variables, which should ignore your build system and automatically forward user flags to the link command line

Yes, this is essentially what I'm trying now, adding the libraries explicitly to the link line for CoinUtils, which is specified with `--with-coinutils-lib`.

jpfeuffer commented:
> Are you sure about this? The library provided by openblas seems to be called `openblas.lib`, not `cblas.lib`. I also installed openblas and didn't get the library.

Ah I see the confusion, I think. I linked the wrong package when formulating my first answer. This is the netlib version on the package explorer:
https://conda-metadata-app.streamlit.app/?q=conda-forge%2Fwin-64%2Flibcblas-3.9.0-9_hd77e832_netlib.conda

 There you should see that cblas.lib is present. If you pin the package on your local machine with e.g. libcblas=*=netlib you should be able to get it.
(In conda-forge CI this is solved by the .ci_support pinnings)

tkralphs committed:
Switch to using arrays

tkralphs commented:
OK, yes, that worked! Now I see `cblas.lib` locally. So then I guess we should also ensure this is installed locally when one installs the coin-or-utils package.

tkralphs committed:
Change ordering of libraries

tkralphs committed:
Trigger build

tkralphs commented:
Weird, that last commit didn't trigger a build for some reason.

tkralphs commented:
OK, now I got the library onto the link line and everything looks correct, just like the CoinUtils link line, but the linking is still failing for some reason. 

[Good Build](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1183419&view=logs&j=a70f640f-cc53-5cd3-6cdc-236a1aa90802&t=6119ccbe-9301-594f-7c27-f792b80a7fcc&l=1938)

[Failing Build](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1183859&view=logs&j=fe7faf59-69bb-5487-b8fd-a35a7bda3a51&t=55731638-dd49-5144-2129-5101b9dd1960&l=1916)

It seems like perhaps a library ordering issue, but I tried different orders to no avail. I can probably just try this build locally now that I know how to get `cblas.lib`.

jpfeuffer commented:
Almost! I think I would try one more time in reverse link order from largest to smallest.
Your objects, libCoinUtils.lib, lapack.lib, cblas.lib.

Unfortunately, I cannot edit your branch, otherwise I would've quickly tried.

tkralphs committed:
Use MKL instead of cblas

https://www.intel.com/content/www/us/en/developer/tools/oneapi/onemkl-link-line-advisor.html

tkralphs committed:
Need to install mkl-static

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I was trying to look for recipes to lint for you, but it appears we have a merge conflict. Please try to merge or rebase with the base branch to resolve this conflict.

Please ping the 'conda-forge/core' team (using the `@` notation in a comment) if you believe this is a bug.

tkralphs committed:
Merge branch 'main' into windows-build-clp

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

tkralphs commented:
After some more digging, I think I finally _really_ solved all the mysteries. I was able to reproduce the error locally. The `cblas.lib` and `lapack.lib` libraries we were trying to link to are not what we want. Examining them with `dumpbin`, they don't define any symbols, they may be some kind export library associated with a DLL, I'm not sure. The CoinUtils unit test links despite this because it apparently does not actually require any Blas/Lapack functions. It turned out that if I just remove the libraries from the link line altogether, the unit test still links. So the fact that CoinUtils was working is a red herring. 

I finally decided to just look at the MKL libraries and it turns out they have an `mkl-static` package that very explicitly provides statically linked libraries. Thanks for the very useful link advisor [here](https://www.intel.com/content/www/us/en/developer/tools/oneapi/onemkl-link-line-advisor.html), I was able to pinpoint exactly what libraries I need to statically link with C linkage and boom, I could link Clp. The build is running now and fingers crossed it will succeed.

tkralphs committed:
Few adjustments

tkralphs commented:
OK, finally! But now we should go back and fix up CoinUtils and Osi. Also, the way this is set up is not ideal, since the library names are hard-coded here and they need to match what is used in the other projects. This is what `pkg-config` was designed for. I want to think a little about whether there is a more robust approach. There is a `pkg-config` file installed with the package, but we disabled `pkg-config` support on Windows some time back because of issues that arose and couldn't be fixed easily. There's also another file that is installed in `share/coin/doc/CoinUtils/coinutils_addlibs.txt` that contains the raw link line passed to `cl` when building CoinUtils. Anyway, let me see.

tkralphs commented:
@jpfeuffer I'm trying to understand if there are any changes to `meta.yml` needed to ensure that `mkl-static` is installed at run-time, since it is potentially needed in order to use the libraries. I guess that for this package to be used in any other way than just running the `clp` executable, `pkg-config` support is important, but it is messy to use `pkg-config` on Windows. But at the very least, do we need to put `mkl-static` in `run_exports` or something. I've read the documentation, but it's still not perfectly clear to me.

jpfeuffer commented:
TLDR: To keep it simple it won't hurt adding it to run-exports. There might be some "overprovisioning" for some use cases but maybe better than a missing dependency.

Your case is complicated. I think ideally you would create sub packages in the meta yaml. One for the exe and one for the lib. I think you will need different build scripts though (or one pre build step that builds everything and then two separate install scripts). It also depends a bit on if mkl-static adds itself to its run-exports. If it does, having it in the host section is enough since it will propagate itself. For a shared library I would always assume this to be true but I am not sure about static libraries since you do not know how they are going to be used. We can check.

tkralphs committed:
Restructure to match other projects

tkralphs committed:
Typos

tkralphs committed:
Fix typo, but is this what is wrong?

tkralphs committed:
Oops, one last typo
